src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -237,6 +244,27 @@ public IEnumerable<char> PotentialCommitCharacters
                 return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.None);
             }
 
+            // Specifically for snippets, we use reflection to try and invoke the LanguageServerSnippetExpander's
+            // TryExpand method and determine if it succeeded or not.
+            if (SnippetCompletionItem.IsSnippet(roslynItem))
+            {
+                change.Properties!.TryGetValue("LSPSnippet", out var lspSnippetText);
+
+                if (!document.TryGetText(out var sourceText))
+                {
+                    FatalError.ReportAndCatch(new InvalidOperationException("Document is not loaded and available."), ErrorSeverity.Critical);
+                    return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.None);
+                }
+
+                if (!_roslynLSPSnippetExpander.TryExpand(change.TextChange, sourceText, lspSnippetText, _textView, triggerSnapshot))
```

Comment:
why do we need sourceText *and* triggerSnapshot?  Note that you can get a SourceText from an ITextSnapshot using .AsText() on it.

Reply:
I did not know that.